src/language/robolang.js

```diff
@@ -0,0 +1,294 @@
+    } else if (this.type === ASTNodeTypes.PROGRAM) {
+      repr += "PROGRAM";
+    } else {
+      repr += "UNKNOWN";
+    }
+
+    if (this.children.length > 0) {
+      repr += "{ ";
+      for (var i = 0; i < this.children.length; ++i) {
+        repr += this.children[i].toString() + " ";
+      }
+      repr += "}";
+    }
+
+    return repr;
+  },
+};
+
+/// Stores the state of the syntatic analyzer during parsing.
+function ParserState(tokens) {
+  this.i = 0;
+  this.tokens = tokens;
+  this.error = null;
+}
+
+ParserState.prototype = {
+  programEnded: function() {
+    return this.i >= this.tokens.length;
+  },
+  consumeToken: function() {
+    var token = this.tokens[this.i];
+    ++this.i;
+    return token;
+  },
+  lookahead: function() {
+    return this.tokens[this.i];
+  },
+};
+
+/// Base class for a parser of one node type.
+function ASTNodeParser() { }
+ASTNodeParser.prototype = {
+  /// Returns true if this parser recognizes the upcoming construct to be one
+  /// of the type it knows how to parse.
+  lookahead: function(parserState) {
+    throw "Not implemented.";
+  },
+  /// Parses the next construct of the program and returns an AST node that
+  /// represents it.
+  parse: function(parserState) {
+    throw "Not implemented.";
+  },
+};
+
+/// List of all specific node parser types.
+ASTNodeParser.nodeParsers = [];
+
+/// Discover the type of AST node adequate to the following construct and parses
+/// it using the appropriate parser.
+function ASTGeneralNodeParser() { }
+ASTGeneralNodeParser.prototype = Object.create(ASTNodeParser.prototype);
+Object.assign(ASTGeneralNodeParser.prototype, {
+  /// Returns true if any of the parsers recognizes the upcoming construct.
+  lookahead: function(parserState) {
+    for (var i = 0; i < ASTNodeParser.nodeParsers.length; ++i) {
+      var parser = new ASTNodeParser.nodeParsers[i]();
+      if (parser.lookahead(parserState)) {
+        return true;
+      }
+    }
+    return false;
+  },
+  /// Finds the appropriate parser and uses it.
+  parse: function(parserState) {
+    for (var i = 0; i < ASTNodeParser.nodeParsers.length; ++i) {
+      var parser = new ASTNodeParser.nodeParsers[i]();
+      if (parser.lookahead(parserState)) {
+        return parser.parse(parserState);
+      }
+    }
+    throw "No suitable parser found.";
+  },
+});
+
+function ASTProgramNodeParser() { }
+ASTProgramNodeParser.prototype = Object.create(ASTNodeParser.prototype);
+Object.assign(ASTProgramNodeParser.prototype, {
+  lookahead: function(parserState) {
+    return parserState.lookahead().type == TokenTypes.BEGIN_BLOCK;
+  },
+
+  parse: function(parserState) {
+    var node = new ASTNode(ASTNodeTypes.PROGRAM);
+    var parser = new ASTGeneralNodeParser();
+
+    while (!parserState.programEnded() && parser.lookahead(parserState)) {
+      node.children.push(parser.parse(parserState));
+    }
+
+    return node;
+  },
+});
+
+ASTNodeParser.nodeParsers.push(ASTBlockNodeParser);
```

Comment:
Because otherwise we'd have left recursion in the grammar. We'd have Program -> Program | ..., but then this recursive descent algorithm would enter an infinite loop. Makes sense?